src/main/scala/io/iohk/ethereum/db/storage/StateStorage.scala

```diff
@@ -0,0 +1,128 @@
+package io.iohk.ethereum.db.storage
+
+import io.iohk.ethereum.db.cache.MapCache
+import io.iohk.ethereum.db.dataSource.DataSource
+import io.iohk.ethereum.db.storage.NodeStorage.{NodeEncoded, NodeHash}
+import io.iohk.ethereum.db.storage.pruning.{ArchivePruning, PruningMode}
+import io.iohk.ethereum.mpt.MptNode
+import io.iohk.ethereum.network.p2p.messages.PV63.MptNodeEncoders._
+
+trait StateStorage {
+  def getBackingStorage(bn: BigInt): MptStorage
+  def getReadOnlyStorage: MptStorage
+
+  def onBlockSave(bn: BigInt, currentBestSavedBlock: BigInt)(updateBestBlock: () => Unit): Unit
+  def onBlockRollback(bn: BigInt, currentBestSavedBlock: BigInt)(updateBestBlock: () => Unit): Unit
+
+  def saveNode(nodeHash: NodeHash, nodeEncoded: NodeEncoded, bn: BigInt)
+  def getNode(nodeHash: NodeHash): Option[MptNode]
+  def forcePersist: Unit
+}
+
+class ArchiveStateStorage(private val nodeStorage: NodeStorage,
+                          private val cachedNodeStorage: CachedNodeStorage) extends StateStorage {
+
+  override def forcePersist: Unit = {
+    cachedNodeStorage.forcePersist()
+  }
+
+  override def onBlockSave(bn: BigInt, currentBestSavedBlock: BigInt)(updateBestBlock: () => Unit): Unit = {
+    if (cachedNodeStorage.persist()) {
+      updateBestBlock()
+    }
+  }
+
+  override def onBlockRollback(bn: BigInt, currentBestSavedBlock: BigInt)(updateBestBlock: () => Unit): Unit = {
+    if (cachedNodeStorage.persist()) {
+      updateBestBlock()
+    }
+  }
+
+  override def getReadOnlyStorage: MptStorage = {
+    new SerializingMptStorage(ReadOnlyNodeStorage(new ArchiveNodeStorage(cachedNodeStorage)))
+  }
+
+  override def getBackingStorage(bn: BigInt): MptStorage = {
+    new SerializingMptStorage(new ArchiveNodeStorage(cachedNodeStorage))
+  }
+
+  override def saveNode(nodeHash: NodeHash, nodeEncoded: NodeEncoded, bn: BigInt): Unit = {
+    nodeStorage.put(nodeHash, nodeEncoded)
+  }
+
+  override def getNode(nodeHash: NodeHash): Option[MptNode] = {
+    cachedNodeStorage.get(nodeHash).map(_.toMptNode)
+  }
+}
+
+class ReferenceCountedStateStorage(private val nodeStorage: NodeStorage,
+                                   private val cachedNodeStorage: CachedNodeStorage,
+                                   private val pruningHistory: BigInt) extends StateStorage {
+  override def forcePersist: Unit = {
+    cachedNodeStorage.forcePersist()
+  }
+
+  override def onBlockSave(bn: BigInt, currentBestSavedBlock: BigInt)(updateBestBlock: () => Unit): Unit = {
+    val blockToPrune = bn - pruningHistory
+
+    if (blockToPrune <= currentBestSavedBlock){
```

Comment:
It could be simplified to `ReferenceCountNodeStorage.prune(blockToPrune, cachedNodeStorage, inMemory = blockToPrune > currentBestSavedBlock)`